### DISCUSSION TITLE
Loading environment variables in nuxt.config

### DISCUSSION THREAD

**Discussion by mirabledictu** [nuxt/nuxt#18329] [Questions]
Hello, how do we read `.env` files from `nuxt.config.ts`?

I tried `process.env` but it doesn't work. I have to add this at the top to make it work

```ts
import { config } from 'dotenv'
config()

export default defineNuxtConfig({ ... }) 
```

Thank you!

**Comment by HiddenUndefined** [comment]
I have a bad, but temporary, solution.

1. Create `env.js` in root directiory.
1.1 If you have `git repository` add `env.js` in `.gitignore`!
2. Modify `nuxt.config.ts`:
```js
// Env
import { envObj } from './env.js'


// Config
export default defineNuxtConfig({
  /**
   * Environment
   */
  // @ts-ignore
  env: envObj,
})
```

But I waiting good solution with `.env` file like in `Nuxt 2` :)

**Reply by wobsoriano** [reply]
yeah you're gonna have problems with that when deploying to serverless platforms

**Reply by HiddenUndefined** [reply]
I know, but I want to find the right solution.

Unfortunately, I could not find the correct solution in the documentation :(

**Reply by wobsoriano** [reply]
His solution works tho

```ts
import { config } from 'dotenv'
config()

export default defineNuxtConfig({ ... }) 
```

**Reply by HiddenUndefined** [reply]
I tried this solution, but I can't get access to the variables in the components `script`.
When I call `proces.env.{ SOMEONE VARIABLES }` – I get `undefined`.

**Reply by Zielgestalt** [reply]
Here someone shared his solution with useRuntimeConfig in script setup: https://twitter.com/alexjoverm/status/1448264523467657225 I haven't try it out, but perhaps that helps.

**Reply by wobsoriano** [reply]
> Here someone shared his solution with useRuntimeConfig in script setup: https://twitter.com/alexjoverm/status/1448264523467657225 I haven't try it out, but perhaps that helps.

This is the right answer

**Reply by HiddenUndefined** [reply]
I think this is a great solution, but it requires you to use Vite, I prefer to abandon it for now and unfortunately, this solution does not work without it...

**Reply by danielroe** [reply]
@GeorgeRudenko `useRuntimeConfig()` shouldn't require you to use vite. What makes you say that?

**Reply by HiddenUndefined** [reply]
It's all my IDEA... it import useRuntimeConfig from nuxt3 and I get many error with mjs & vite :)

**Reply by danielroe** [reply]
Yes, this is how to use it:
```js
// this line is optional if you're using auto-imports
import { useRuntimeConfig } from '#app'

const config = useRuntimeConfig()
```

If you're struggling with IDE, make sure your `tsconfig.json` contains the following:
```json
{
  "extends": "./.nuxt/tsconfig.json"
}
```

**Reply by wobsoriano** [reply]
so for loading variables in nuxt config, we need to load it manually?

**Reply by danielroe** [reply]
... extracting to a top-level answer for clarity

**Comment by danielroe** [comment]
You shouldn't need to run `dotenv` manually; that would be a bug (follow at https://github.com/nuxt/framework/issues/1085)

To define runtime config:
```js
import { defineNuxtConfig } from 'nuxt3';

export default defineNuxtConfig({
  publicRuntimeConfig: {
    test: process.env.TEST_VAL || 'nothing provided',
  },
  privateRuntimeConfig: {
    secretToken: process.env.TOKEN,
  }
}
```

To access runtime config within your Nuxt app:

```js
// this line is optional if you're using auto-imports
import { useRuntimeConfig } from '#app'

const config = useRuntimeConfig()
// this is only defined on server
config.secretToken
// this is always accessible
config.test
```

If you're struggling with IDE, make sure your `tsconfig.json` contains the following:
```json
{
  "extends": "./.nuxt/tsconfig.json"
}
```
